File: J9Ch/src/J_7_Two_Pointers/Required_10/_609_Two_Sum_Less_than_or_equal_to_target_Medium.java
package J_7_Two_Pointers.Required_10;
import java.util.*;
/** 609 Two Sum - Less than or equal to target
 * Medium

 * Created by tianhuizhu on 6/28/17.
 */
public class _609_Two_Sum_Less_than_or_equal_to_target_Medium {

    public class Solution {
        /**
         * @param nums an array of integer
         * @param target an integer
         * @return an integer
         */
        public int twoSum5(int[] nums, int target) {
            // Write your code here
            if (nums == null || nums.length < 2)
                return 0;

            Arrays.sort(nums);
            int cnt = 0;
            int left = 0, right = nums.length - 1;
            while (left < right) {
                int v = nums[left] + nums[right];
                if (v > target) {
                    right --;
                } else {
                    cnt += right - left;
                    left ++;
                }
            }
            return cnt;
        }
    }
}

File: J9Ch/src/J_7_Two_Pointers/Required_10/python_cpp/_609_Two_Sum_Less_than_or_equal_to_target_Medium.cpp
/** 609 Two Sum - Less than or equal to target
 * Medium*/


class Solution {
public:
    /**
     * @param nums an array of integer
     * @param target an integer
     * @return an integer
     */
    int twoSum5(vector<int> &nums, int target) {
        // Write your code here
        int n = nums.size(), cnt = 0;
        sort(nums.begin(), nums.end());
        int l = 0, r = nums.size() -1;
        while (l < r){
            if (nums[l] + nums[r] > target){
                r--;
            } else {
                cnt += r - l;
                l++;
            }
        }
        return cnt;
    }
};